System.IO.FileSystem/Directory.cs

```diff
@@ -3,151 +3,299 @@
+                    ((attributes & (uint)FileAttributes.ReadOnly) != 0))
+                {
+                    // it's readonly or not a directory
+                    throw new IOException(
+                        string.Empty,
+                        (int)IOException.IOExceptionErrorCode.UnauthorizedAccess);
+                }
+
+                // make sure it is indeed a directory (and not a file)
+                if (!Exists(path))
+                {
+                    throw new IOException(
+                        string.Empty,
+                        (int)IOException.IOExceptionErrorCode.DirectoryNotFound);
+                }
+
+                NativeIO.Delete(
+                    path,
+                    recursive);
+            }
+            finally
+            {
+                // regardless of what happened, we need to release the directory when we're done
+                FileSystemManager.UnlockDirectory(record);
+            }
         }
 
         /// <summary>
-        /// Determines whether the specified directory exists.
+        /// Determines whether the given path refers to an existing directory on disk.
         /// </summary>
-        /// <param name="path">Path to the directory.</param>
-        /// <returns>True if directory under given path exists, otherwise it returns false.</returns>
+        /// <param name="path">The path to test.</param>
+        /// <returns><see langword="true"/> if <paramref name="path"/> refers to an existing directory; <see langword="false"/> if the directory does not exist or an error occurs when trying to determine if the specified directory exists.</returns>
         /// <exception cref="ArgumentNullException">Path must be defined.</exception>
         /// <exception cref="IOException">Invalid drive or path to the parent folder doesn't exist.</exception>
         public static bool Exists(string path)
         {
-            return ExistsNative(path);
+            // path validation happening in the call
+            path = Path.GetFullPath(path);
+
+            // is this the absolute root? this always exists.
+            if (path == NativeIO.FSRoot)
+            {
+                return true;
+            }
+            else
+            {
+                try
+                {
+                    uint attributes = NativeIO.GetAttributes(path);
+
+                    if (attributes == NativeIO.EmptyAttribute)
+                    {
+                        // this means not found
+                        return false;
+                    }
+
+                    if ((((FileAttributes)attributes)
+                         & FileAttributes.Directory) == FileAttributes.Directory)
+                    {
+                        // this is a directory
+                        return true;
+                    }
+                }
+                catch (Exception)
+                {
+                    return false;

```

Comment:
and you can remove this one as it returns at the end

Reply:
Actually it's more than that: the try/catch is not doing anything usefull...

Reply:
then if nothing can fail, you can simplify even more :-D